### PR TITLE
Vitis-3414 xbflash2: enhance xbflash tool command line options

### PR DESCRIPTION
**Pull request by rajkumar-xilinx**
Moved and updated xbflash2 tool to make it production ready.

New package named xrt_<>-xbflash2.deb will be created along with xrt*.deb. It has to be installed in order to use xbflash2 tool.

Signed-off-by: Rajkumar Rampelli <rajkumar@xilinx.com>

### CODE REVIEW

**Comment by chvamshi-xilinx** [comment]
Open it again once all changes are done

**Comment by rajkumar-xilinx** [comment]
retest this please

**Comment by rajkumar-xilinx** [comment]
**On U2**
root@rahul:/home/rahul/raj/XRT/build/Release/usr/local/bin# ./xbflash2 program --flash spi --image /lib/firmware/xilinx/625b99fa75b56d8353ff2a7a999c8bbb/partition.mcs --device 0000:62:00.0 --image /lib/firmware/xilinx/625b99fa75b56d8353ff2a7a999c8bbb/partition.mcs
About to flash below MCS bitstream onto device 0000:62:00.0:
Are you sure you wish to proceed? [y/n]: y
Successfully opened /dev/xfpga/flash.m25088.0
flashing via QSPI driver
Bitstream guard installed on both flashes @0x801000
Extracting bitstream from MCS data:
.........
Extracted 9245382 bytes from bitstream @0x801000
Writing bitstream to flash 0:
.........
Extracting bitstream from MCS data:
.........
Extracted 9245382 bytes from bitstream @0x801000
Writing bitstream to flash 1:
.........
Bitstream guard removed from both flashes
root@rahul:/home/rahul/raj/XRT/build/Release/usr/local/bin#

**Review by chvamshi-xilinx — CHANGES_REQUESTED** [review]
Disable it for edge also. Looks like build is failing for edge.

File: src/runtime_src/core/tools/CMakeLists.txt [hunk 1]
```diff
@@ -10,6 +10,10 @@ if (${XRT_NATIVE_BUILD} STREQUAL "yes")
   add_subdirectory(xbmgmt2)
 endif()
 
+if(NOT XRT_EDGE)
```
Comment (chvamshi-xilinx) [inline src/runtime_src/core/tools/CMakeLists.txt:13]: Please use XRT_NATIVE_BUILD flag for consistency. look at xbmgmt2

**Review by AShivangi — APPROVED** [review]

**Comment by dayeh-xilinx** [comment]
retest this please.

**Review by rozumx — CHANGES_REQUESTED** [review]
Unfortunately, I only got through half of the code regarding this code review and had to stop.

Take aways:
1. The code is base upon C coding styles and not C++.  For example:
  a. Return error codes instead of throwing exceptions.
  b. Pass by pointers instead of pass by reference.
  c. Repeating common functionality between methods 
  d. Multiple exit points.
2. I understand that this code was based upon earlier code, unfortunately it wasn't cleaned up.  It should be.
3. Common helper code in core/tools/common is not being used.  This code has a lot of duplicate functionality.
4. Compiler OS directives are being used.  This should not occur of any code in core/tools, with the exception of pragmas to turn off some compiler warnings.
5. Helper code to report the boost program_options is not being used for help like functionality.

File: src/runtime_src/core/tools/xbflash2/cmd_dump.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      if (!soffset.empty()) {
+        std::stringstream sstream(soffset);
+        sstream >> offset;
+      }
+      slength = vm["length"].as<std::string>();
+      if (!slength.empty()) {
+        std::stringstream sstream(slength);
+        sstream >> len;
+      }
+    } catch (...) {
```
[the remark this answers is not in the document]
Comment (rajkumar-xilinx) [reply src/runtime_src/core/tools/xbflash2/cmd_dump.cpp:73]: these are optional arguments to the tool, so we are not returning any error if these args are not found in the command line.

File: src/runtime_src/core/tools/xbflash2/firmware_image.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+#endif
+
+firmwareImage::firmwareImage(const char *file) :
+    mBuf(nullptr)
+{
+    std::ifstream in(file, std::ios::binary | std::ios::ate);
+    if (!in.is_open())
+    {
+        this->setstate(failbit);
+        std::cout << "Can't open " << file << std::endl;
```
[the remark this answers is not in the document]
Comment (rajkumar-xilinx) [reply src/runtime_src/core/tools/xbflash2/firmware_image.cpp:31]: code sets state to "fail" and caller function will check the state before proceeding.

**caller function code:**

firmwareImage pri(primary_file[0].c_str());
if (pri.fail())
      return -EINVAL;

File: src/runtime_src/core/tools/xbflash2/main.cpp [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+#ifndef _WIN32
+    const char* SudoMessage = "ERROR: root privileges required.";
+    if ((getuid() == 0) || (geteuid() == 0))
+      return;
+    std::cout << SudoMessage << std::endl;
+    exit(-EPERM);
+#endif
+}
+
+bool canProceed()
```
[the remark this answers is not in the document]
Comment (rajkumar-xilinx) [reply src/runtime_src/core/tools/xbflash2/main.cpp:60]: xbflash2 may not reuse these functionality due to xrt core lib dependency. This tool has to be work without xrt.

File: src/runtime_src/core/tools/xbflash2/main.cpp [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+      answered = (input.compare("y") == 0 || input.compare("n") == 0);
+    }
+
+    proceed = (input.compare("y") == 0);
+    if (!proceed)
+      std::cout << "Action canceled." << std::endl;
+    return proceed;
+}
+
+void printSubCmdHelp(const std::string& subCmd)
```
[the remark this answers is not in the document]
Comment (rajkumar-xilinx) [reply src/runtime_src/core/tools/xbflash2/main.cpp:78]: xbflash2 may not reuse these functionality due to xrt core lib dependency. This tool has to be work without xrt.

File: src/runtime_src/core/tools/xbflash2/cmd_dump.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+#include <map>
+#include <functional>
+
+#ifdef _WIN32
+# pragma warning( disable : 4267)
+#endif
+
+// For backward compatibility
+const char *subCmdDumpDesc = "Reads the image(s) for a given device for a given length and outputs the same to given file. It is applicable for only QSPIPS flash.";
+const char *subCmdDumpUsage =
```
[the remark this answers is not in the document]
Comment (rajkumar-xilinx) [reply src/runtime_src/core/tools/xbflash2/cmd_dump.cpp:31]: I have not explored it actually as I was told enhancement scope of this tool is very limited in future. So, I mostly copied from old code.

File: src/runtime_src/core/tools/xbflash2/cmd_program.cpp [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+      if (!sBar.empty())
+        bar = std::stoi(sBar);
+      sBarOffset = vm["bar-offset"].as<std::string>();
+      if (!sBarOffset.empty()) {
+        std::stringstream sstream(sBarOffset);
+        sstream >> baroff;
+      }
+    } catch (...) {
+    }
+
```
[the remark this answers is not in the document]
Comment (rajkumar-xilinx) [reply src/runtime_src/core/tools/xbflash2/cmd_program.cpp:295]: Will add check to read if the offset is valid and assign it to offset. No need to return error as it is an optional command line arg.

**Comment by rajkumar-xilinx** [comment]
Hi @rozumx , thank you for reviewing the PR.

>>>What is this code here. It excludes windows. It should be refactor down to a common API that is implemented in Linux and windows.
Work is in progress to refactor the code to support windows API as well. But, it will take time to complete it. So, we are going to enable this feature on linux for now

**Comment by dayeh-xilinx** [comment]
retest this please. Cancelled from queue to give way to PR#6149

**Comment by chvamshi-xilinx** [comment]
Naresh will send a new PR with suggested changes.

**Pull request closed without merging**